### PR TITLE
Wu2021rethinking 2

### PR DESCRIPTION
**Pull request by aabayomi**

### CODE REVIEW

File: _summaries/wu2021rethinking_2.md [hunk 1]
```diff
@@ -1,19 +1,90 @@
...
 TODO: Summarize the paper:
-* What is the core idea?
-* How is it realized (technically)?
-* How well does the paper perform?
-* What interesting variants are explored?
+
+### What is the core idea?
+
+The authors argues using positive random orthogonal features can estimate softmax showing speed and performance gains over SOTA transformer architecture
+
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:15]: Can you also add how the paper changes attention computation for quadratic time complexity to linear time complexity?

File: _summaries/wu2021rethinking_2.md [hunk 2]
```diff
@@ -1,19 +1,90 @@
...
+where Λ = exp(−∥x∥2+∥y∥2 ) and cosh is hyperbolic cosine. Consequently, softmax-kernel admits a
+ 2
+positive random feature map unbiased approximation with h(x) = exp(− ∥x∥2 ), l = 1, f1 = exp 2
+and D = N(0,Id) or: h(x) = √1 exp(−∥x∥2 ), l = 2, f1(u) = exp(u), f2(u) = exp(−u) and the 22
++
+􏱂m same D (the latter for further variance reduction). We call related estimators: SM
+hyp+
+􏱂m
+``` -->
+
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:35]: The paper also discusses on why non-negative kernels are useful. Can you please add that?

File: _summaries/wu2021rethinking_2.md (hunk 1 above)
Comment (DartingMelody) [reply _summaries/wu2021rethinking_2.md:15]: Also, can you add that paper introduced Performer which uses linear space and complexity without depending on priors like low-rankness or sparsity? Also can you talk a bit about FAVOR+.

File: _summaries/wu2021rethinking_2.md [hunk 3]
```diff
@@ -1,19 +1,90 @@
...
+[2] Use of orthogonal random features (ORFs)
+
+Well known method to reduce the variance of the estimator  by selecting random features and orthogonal random samples.
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:38]: It's kind of difficult to follow this line, maybe you can give some context.  Also, add about the representational power of FAVOR+ on large scale tasks.

File: _summaries/wu2021rethinking_2.md [hunk 4]
```diff
@@ -1,19 +1,90 @@
...
+􏱂m same D (the latter for further variance reduction). We call related estimators: SM
+hyp+
+􏱂m
+``` -->
+
+[2] Use of orthogonal random features (ORFs)
+
+Well known method to reduce the variance of the estimator  by selecting random features and orthogonal random samples. 
+
+FAVOR+(Performer) = ORFs + PSFs
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:40]: PSFs - can you add the full form and also a one-line explanation of this line here.

File: _summaries/wu2021rethinking_2.md [hunk 5]
```diff
@@ -1,19 +1,90 @@
...
+
+
+
+### How well does the paper perform?
+
+<img width="700px" src="wu2021rethinking_2b.png"/>
+
+<img width="700px" src="wu2021rethinking_2g.png"/>
+
+Baseline comparison  of FAVOR+ (Performer) with normal Transformer models for both unidirectional/causal modeling and bidirectional/masked language modeling tasks
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:52]: Is this the caption of the previous image ?

File: _summaries/wu2021rethinking_2.md [hunk 6]
```diff
@@ -1,19 +1,90 @@
...
+
+- BookCorpus (Zhu et al., 2015)  
+
+- Wikipedia dataset used in BERT (Devlin et al., 2018)
+
+
+
+<img width="700px" src="wu2021rethinking_2c.png"/>
+
+Orthogonal features show lower softmax attention approximation error. Also using positive features produced lower errors compared to $sine/cosine$ features.
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:66]: You missed $sine/cosine$  to $$sine/cosine$$

File: _summaries/wu2021rethinking_2.md [hunk 7]
```diff
@@ -1,19 +1,90 @@
...
+
+
+<img width="700px" src="wu2021rethinking_2d.png"/>
+
+Showed backward compatibility with pre-trained models benefitting from softmax approximation.
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:71]: Did you mean performers displayed backward compatibility with pre-trained models benefitting from softmax approximation.

File: _summaries/wu2021rethinking_2.md [hunk 8]
```diff
@@ -1,19 +1,90 @@
...
+
+<img width="700px" src="wu2021rethinking_2e.png"/>
+
+Performer combined with a ReLU showed better training performance on a 36-layer model  on TrEMBL protein sequence dataset (Consortium, 2019). 
+
+
+<img width="700px" src="wu2021rethinking_2f.png"/>
+
+Performer showed better training performance on ImageNet64 dataset.
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:80]: in comparison to which models (Are you referring to Reformer, Linformer)

File: _summaries/wu2021rethinking_2.md [hunk 9]
```diff
@@ -1,19 +1,90 @@
...
+### What interesting variants are explored?
 
 ## TL;DR
-* Three
-* Bullets
-* To highlight the core concepts
+
+* Choosing Positive random features are better than standard trigonometric features.
+
+* Transformers requires a huge cost of computation but Performer are more efficient.
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:88]: you can also add peformer uses FAVOR+ provides time and space-efficient approximation of attention.

File: _summaries/wu2021rethinking_2.md [hunk 10]
```diff
@@ -1,19 +1,90 @@
...
+
+### What interesting variants are explored?
 
 ## TL;DR
-* Three
-* Bullets
-* To highlight the core concepts
+
+* Choosing Positive random features are better than standard trigonometric features.
+
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:87]: Another point could be performers are fully compatible with Transfomer and provide strong theoretical guarantees, without any priors.

**Review by DartingMelody — CHANGES_REQUESTED** [review]
Hi, Your summary has the essential points but it is somewhat difficult to follow. Can you please explain a bit more and have a flow in the summary. That will improve your summary a lot. Let me know if you have any doubts or need any help.

File: _summaries/wu2021rethinking_2.md [hunk 11]
```diff
@@ -1,19 +1,90 @@
...
+### What is the core idea?
+
+The authors argues using positive random orthogonal features can estimate softmax showing speed and performance gains over SOTA transformer architecture
+
+### How is it realized (technically)?
+
+[1] Only non-negative kernel scores are useful.
+
+<img width="700px" src="wu2021rethinking_2h.png"/>
+
```
Comment (DartingMelody) [inline _summaries/wu2021rethinking_2.md:21]: Will you be explaining this, while presenting or else you can replace this picture with more content on why non-negative kernel scores are useful. How they have technically reduced the time complexity of attention from O(L^2 * D) to O(Lrd). Also maybe add just the required math expressions.

File: _summaries/wu2021rethinking_2.md (hunk 4 above)
Comment (DartingMelody) [reply _summaries/wu2021rethinking_2.md:40]: And also explain the algorithm below in short.

**Comment by aabayomi** [comment]
Thank you, I have made the reviewed changes. let me know if there is anything else missing.

**Comment by zhaoyue-zephyrus** [comment]
@aabayomi could you do me a favor by removing changes on `reddi2019_1` and rename all `wu2021rethinking_XX`s to `choromanski2021rethinking_XX` otherwise there are conflicts with previous reviews.

**Comment by aabayomi** [comment]
> @aabayomi could you do me a favor by removing changes on `reddi2019_1` and rename all `wu2021rethinking_XX`s to `choromanski2021rethinking_XX` otherwise there are conflicts with previous reviews? You can open a new PR if you want.

**Comment by aabayomi** [comment]
@zhaoyue-zephyrus  sorry for this conflict by removing do you mean delete?

**Comment by zhaoyue-zephyrus** [comment]
@aabayomi you can start from the latest main branch, checkout a new branch, and add only `choromanski2021rethinking_2`-related files in it, and initiate a new PR.

**Comment by zhaoyue-zephyrus** [comment]
closing this because of a newer PR #109
